File: kotlin/src/main/kotlin/v1/CommsSDK.kt
package v1

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.web.client.RestTemplate
import v1.models.*
import v1.utils.*

final class CommsSDK {
    var userName: String = ""
        private set
    var apiKey: String = ""
        private set
    var isAuthenticated: Boolean = false
        private set
    var senderId: String = "EgoSMS"
        private set

    companion object {
        var API_URL = "https://comms.egosms.co/api/v1/json/"
        val OBJECT_MAPPER = ObjectMapper()
        val client = RestTemplate()

        fun authenticate(userName: String, apiKey: String): CommsSDK {
            val commsSDK = CommsSDK()
            commsSDK.userName = userName
            commsSDK.apiKey = apiKey
            commsSDK.isAuthenticated = Validator.validateCredentials(commsSDK)
            return commsSDK
        }

        /**
         * Uses the sandbox url - useful for testing scenarios.
         * <br></br>
         * Make an account at "[comms-test.pahappa.net](https://comms-test.pahappa.net)" to use the sandbox.
         * Use [useLiveServer] for the live server.
         */
        fun useSandBox() {
            API_URL = "https://comms-test.pahappa.net/api/v1/json"
        }

        /**
         * Uses the live url - useful for actual messaging scenarios.
         * <br></br>
         * Make an account at "[comms.egosms.co](https://comms.egosms.co)" to use the live server.
         * Use [useSandBox] for the sandbox server.
         */
        fun useLiveServer() {
            API_URL = "https://comms.egosms.co/api/v1/json"
        }

    }

    fun withSenderId(senderId: String): CommsSDK {
        this.senderId = senderId
        return this
    }

    fun sendSMS(
        number: String,
        message: String,
        senderId: String = this.senderId,
        priority: MessagePriority = MessagePriority.HIGH
    ): Boolean {
        return sendSMS(listOf(number), message, senderId, priority)
    }

    fun sendSMS(
        numbers: List<String>,
        message: String,
        senderId: String = this.senderId,
        priority: MessagePriority = MessagePriority.HIGH
    ): Boolean {
        val apiResponse = querySendSMS(numbers, message, senderId, priority)
        if (apiResponse == null) {
            println("Failed to get a response from the server.")
            return false
        }
        when (apiResponse.status) {
            ApiResponseCode.OK -> {
                println("SMS sent successfully.")
                println("MessageFollowUpUniqueCode: " + apiResponse.messageFollowUpCode)
                return true
            }

            ApiResponseCode.Failed -> {
                println("Failed: ${apiResponse.message}")
                return false
            }
            else -> throw RuntimeException("Unexpected response status: " + apiResponse.status)
        }
    }

    /** Same as [sendSMS] but returns the full [ApiResponse] object. */
    fun querySendSMS(
        number: String,
        message: String,
        senderId: String = this.senderId,
        priority: MessagePriority = MessagePriority.HIGH
    ): ApiResponse? {
        return querySendSMS(listOf(number), message, senderId, priority)
    }

    /** Same as [sendSMS] but returns the full [ApiResponse] object. */
    fun querySendSMS(
        numbers: List<String>,
        message: String,
        senderId: String = this.senderId,
        priority: MessagePriority = MessagePriority.HIGH
    ): ApiResponse? {
        var numbers = numbers
        var senderId = senderId
        if (sdkNotAuthenticated()) return null
        require(!(numbers.isEmpty())) { "Numbers list cannot be empty" }
        require(!(message.isEmpty())) { "Message cannot be empty" }
        require(message.length != 1) { "Message cannot be a single character" }
        if (senderId.trim { it <= ' ' }.isEmpty()) {
            senderId = this.senderId
        }
        if (senderId.length > 11) {
            println("Warning: Sender ID length exceeds 11 characters. Some networks may truncate or reject messages.")
        }
        numbers = NumberValidator.validateNumbers(numbers)
        if (numbers.isEmpty()) {
            System.err.println("No valid phone numbers provided. Please check inputs.")
            return null
        }
        val apiRequest = ApiRequest()
        apiRequest.method = "SendSms"
        val messageModels: MutableList<MessageModel> = ArrayList()
        for (num in numbers) {
            val messageModel = MessageModel()
            messageModel.number = num
            messageModel.message = message
            messageModel.senderId = senderId
            messageModel.priority = priority
            messageModels.add(messageModel)
        }
        apiRequest.messageData = messageModels
        apiRequest.userdata = UserData(userName, apiKey)
        apiRequest.walletType = WalletType.LOCAL
        val res = client.postForEntity(API_URL, apiRequest, String::class.java)

        try {
            return OBJECT_MAPPER.readValue(res.getBody(), ApiResponse::class.java)
        } catch (e: Exception) {
            System.err.println("Failed to send SMS: " + e.message)
            try {
                System.err.println("Request: " + OBJECT_MAPPER.writeValueAsString(apiRequest))
            } catch (_: Exception) {
            }
            return null
        }
    }

    private fun sdkNotAuthenticated(): Boolean {
        if (!isAuthenticated) {
            System.err.println("SDK is not authenticated. Please authenticate before performing actions.")
            System.err.println("Attempting to re-authenticate with provided credentials...")
            isAuthenticated = Validator.validateCredentials(this);
            return !isAuthenticated;
        }
        return false
    }

    /** Same as [getBalance] but returns the full [ApiResponse] object. */
    fun queryBalance(walletType: WalletType = WalletType.LOCAL): ApiResponse? {
        if (sdkNotAuthenticated()) {
            return null
        }
        val apiRequest = ApiRequest()
        apiRequest.method = "Balance"
        apiRequest.userdata = UserData(userName, apiKey)
        apiRequest.walletType = walletType
        try {
            val res = client.postForEntity(API_URL, apiRequest, String::class.java)
            val response = OBJECT_MAPPER.readValue(res.getBody(), ApiResponse::class.java)
            return response
        } catch (e: Exception) {
            throw RuntimeException("Failed to get balance: " + e.message, e)
        }
    }

    fun getBalance(walletType: WalletType = WalletType.LOCAL): Double? {
        val response = queryBalance(walletType)
        return response?.balance
    }

    override fun toString(): String {
        return "SDK(${this.userName} => ${this.apiKey})"
    }
}

File: js/src/v1/CommsSDK.ts
import { ApiRequest } from "./models/ApiRequest";
import { ApiResponse } from "./models/ApiResponse";
import { MessageModel } from "./models/MessageModel";
import { MessagePriority } from "./models/MessagePriority";
import { UserData } from "./models/UserData";
import { NumberValidator } from "./utils/NumberValidator";
import { Validator } from "./utils/Validator";
import axios from "axios";

export class CommsSDK {
  public static API_URL = "https://comms.egosms.co/api/v1/json/";

  private _apiKey?: string;
  private _userName?: string;
  private _senderId = "EgoSMS";
  private _isAuthenticated = false;

  private constructor() {}

  public static useSandBox(): void {
    CommsSDK.API_URL = "https://comms-test.pahappa.net/api/v1/json";
  }

  public static useLiveServer(): void {
    CommsSDK.API_URL = "https://comms.egosms.co/api/v1/json";
  }

  public static authenticate(userName: string, apiKey: string): CommsSDK {
    const sdk = new CommsSDK();
    sdk._userName = userName;
    sdk._apiKey = apiKey;
    Validator.validateCredentials(sdk);
    return sdk;
  }

  public withSenderId(senderId: string): CommsSDK {
    this._senderId = senderId;
    return this;
  }

  public setAuthenticated(): void {
    this._isAuthenticated = true;
  }

  public get userName(): string {
    return this._userName || "";
  }

  public get apiKey(): string {
    return this._apiKey || "";
  }

  public get isAuthenticated(): boolean {
    return this._isAuthenticated;
  }

  public get senderId(): string {
    return this._senderId;
  }

  // public async sendSMS(
  //     number: string,
  //     message: string,
  //     senderId: string = this._senderId,
  //     priority: MessagePriority = MessagePriority.HIGHEST
  // ): Promise<boolean>;
  // public async sendSMS(
  //     numbers: string[],
  //     message: string,
  //     senderId: string = this._senderId,
  //     priority: MessagePriority = MessagePriority.HIGHEST
  // ): Promise<boolean>;
  public async sendSMS(
    numbers: string | string[],
    message: string,
    senderId: string = this._senderId,
    priority: MessagePriority = MessagePriority.HIGHEST,
  ): Promise<boolean> {
    const numbersArray = Array.isArray(numbers) ? numbers : [numbers];
    const apiResponse = await this.querySendSMS(
      numbersArray,
      message,
      senderId,
      priority,
    );

    if (apiResponse === null) {
      console.log("Failed to get a response from the server.");
      return false;
    }

    if (apiResponse.Status === "OK") {
      console.log("SMS sent successfully.");
      console.log(
        `MessageFollowUpUniqueCode: ${apiResponse.MsgFollowUpUniqueCode}`,
      );
      return true;
    } else if (apiResponse.Status === "Failed") {
      console.log(`Failed: ${apiResponse.Message}`);
      return false;
    } else {
      throw new Error(`Unexpected response status: ${apiResponse.Status}`);
    }
  }

  public async querySendSMS(
    numbers: string[],
    message: string,
    senderId: string,
    priority: MessagePriority,
  ): Promise<ApiResponse | null> {
    if (await this.sdkNotAuthenticated()) {
      return null;
    }
    if (!numbers || numbers.length === 0) {
      throw new Error("Numbers list cannot be empty");
    }
    if (!message) {
      throw new Error("Message cannot be empty");
    }
    if (message.length === 1) {
      throw new Error("Message cannot be a single character");
    }
    if (!senderId || senderId.trim() === "") {
      senderId = this._senderId;
    }

    if (senderId && senderId.length > 11) {
      console.log(
        "Warning: Sender ID length exceeds 11 characters. Some networks may truncate or reject messages.",
      );
    }

    const validatedNumbers = NumberValidator.validateNumbers(numbers);

    if (validatedNumbers.length === 0) {
      console.error("No valid phone numbers provided. Please check inputs.");
      return null;
    }

    const apiRequest = new ApiRequest();
    apiRequest.setMethod("SendSms");

    const messageModels = validatedNumbers.map((number) => {
      const messageModel = new MessageModel();
      messageModel.setNumber(number);
      messageModel.setMessage(message);
      messageModel.setSenderId(senderId);
      messageModel.setPriority(priority);
      return messageModel;
    });

    apiRequest.setMessageData(messageModels);
    apiRequest.setUserdata(new UserData(this._userName!, this._apiKey!));

    try {
      const response = await axios.post(CommsSDK.API_URL, apiRequest.toArray());
      return response.data as ApiResponse;
    } catch (e) {
      console.error(`Failed to send SMS: ${(e as Error).message}`);
      try {
        console.error(`Request: ${JSON.stringify(apiRequest.toArray())}`);
      } catch (_) {
        // Ignore serialization errors
      }
      return null;
    }
  }

  private async sdkNotAuthenticated(): Promise<boolean> {
    if (!this._isAuthenticated) {
      console.error(
        "SDK is not authenticated. Please authenticate before performing actions.",
      );
      console.error(
        "Attempting to re-authenticate with provided credentials...",
      );
      return !(await Validator.validateCredentials(this));
    }
    return false;
  }

  public async queryBalance(): Promise<ApiResponse | null> {
    if (await this.sdkNotAuthenticated()) {
      return null;
    }

    const apiRequest = new ApiRequest();
    apiRequest.setMethod("Balance");
    apiRequest.setUserdata(new UserData(this._userName!, this._apiKey!));

    try {
      const response = await axios.post(CommsSDK.API_URL, apiRequest.toArray());
      return response.data as ApiResponse;
    } catch (e) {
      throw new Error(`Failed to get balance: ${(e as Error).message}`);
    }
  }

  public async getBalance(): Promise<number | null> {
    const response = await this.queryBalance();
    return response?.Balance ?? null;
  }

  public toString(): string {
    return `SDK(${this._userName} => ${this._apiKey})`;
  }
}
